c7n/resources/s3.py

```diff
@@ -1372,6 +1372,166 @@ def process_bucket(self, bucket):
+    `state`: Optional: Defaults to enable
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: s3-dont-ignore-public-acls
+                resource: s3
+                filters:
+                  - type: check-public-block
+                    scope: BlockPublicAcls
+                    enabled: False
+                actions:
+                  - type: set-public-block
+                    # scope: <------ optional (All by default)
+                    #   - BlockPublicAcls
+                    #   - IgnorePublicAcls
+                    # state: enable <------ optional (enable by default)
+    """
+
+    schema = type_schema(
+        'set-public-block',
+        scope={
+            'type': 'array',
+            'items': {'type': 'string',
+                     'enum': ['BlockPublicAcls', 'IgnorePublicAcls',
+                        'BlockPublicPolicy', 'RestrictPublicBuckets', 'All']}},
+        state={'type': 'string',
+               'enum': ['enable', 'disable']})
+    permissions = ("s3:GetBucketPublicAccessBlock", "s3:PutBucketPublicAccessBlock")
+
+    def process(self, buckets):
+        with self.executor_factory(max_workers=3) as w:
+            futures = {w.submit(self.process_bucket, bucket): bucket for bucket in buckets}
+            for future in as_completed(futures):
+                if future.exception():
+                    raise future.exception()
+
+    def process_bucket(self, bucket):
+        s3 = bucket_client(local_session(self.manager.session_factory), bucket)
+        state = self.data.get('state', 'enable')
+        scopes = self.data.get('scope', ['All'])
+        try:
+            config = s3.get_public_access_block(
+                Bucket=bucket['Name'])['PublicAccessBlockConfiguration']
+        except ClientError as e:
+            if e.response['Error']['Code'] == 'NoSuchPublicAccessBlockConfiguration':
+                # Set config to none because NoSuchPublicAccessBlockConfiguration error
+                # was returned. This is important later because it is symbolic of
+                # all blocks being set to false. See if/else statement below
+                config = None
+            else:
+                raise
+        if config:
+            if 'All' in scopes:
+                for key in config.keys():
+                    config[key] = True if state == 'enable' else False
+            else:
+                for scope in scopes:
+                    config[scope] = True if state == 'enable' else False
+            s3.put_public_access_block(
+                Bucket=bucket['Name'],
+                PublicAccessBlockConfiguration=config
+            )
+        else:
+            s3.put_public_access_block(
+                Bucket=bucket['Name'],
+                PublicAccessBlockConfiguration={
+                    'BlockPublicAcls': True if state == 'enable' else False,
+                    'IgnorePublicAcls': True if state == 'enable' else False,
```

Comment:
much of the logic in this method could have been simplified by just defining state as a boolean upfront, it would avoid the 6 duplicates conditions spread throughout.

Reply:
Agreed! I went back with a fresh mind and made a refactor in #5520. I think I was making this much more complicated by using the absent/present enums vs boolean here.